src/devices/ant_antplus.c

```diff
@@ -0,0 +1,164 @@
+
+This decoder only captures and displays the low level packets, identifying the
+2 byte network key plus all other device characteristics and 8 byte ANT payload.
+It identifies ANT+ packets using the unique network key used (0xa6c5)
+Refer to ANT+ documentation for ech specific device to parse the content of
+the pages sent as 8 byte ANT+ payload.
+
+The ANT protocol is using an uncommon strategy for the premable: either 0x55 or
+0xaa, depending on the value of the first bit of the following byte (0 and 1
+respectively). The nRF24L01, on which the ANT protocol is based, uses the same
+premable strategy. In order to determine if the packet is using 0x55 or 0xaa,
+both packets are extracted and the only valid one is determined by matching
+the packet CRC with a calculated CRC value for both alternative packets.
+
+The payload is 18 bytes long structured as follows:
+
+    PNNDDTXLPPPPPPPPCC
+
+- P: Preamble: either 0x55 or 0xAA, depending on the value of first bit of the next byte
+- N: Network key, assume LSB first (ANT+ uses 0xc5a6, most invalid keys 0x255b)
+- D: Device number, 16 bit. LSB first
+- X: Transmission type
+- L: ANT payload length including CRC
+- P: 8 byte ANT or ANT+ payload
+- C: 16 bit CRC (CRC-16/CCITT-FALSE)
+
+Weirdly, L is always 10, because at the moment the ANT payload is always 8 bytes
+
+CREDITS:
+https://github.com/sghctoma/antfs-poc-defcon24
+https://reveng.sourceforge.io/ to reverse engineer the CRC algorithm used
+*/
+
+#include "decoder.h"
+#include "util.h"
+#include "stdbool.h"
+
+static int ant_antplus_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    uint8_t const preamble[] = {0xAA};
+    uint8_t b[17]; // aligned packet data for both preambles/offsets
+    unsigned bit_offset;
+    char payload[8 * 3 + 1]; // payload is 8 hex pairs for ANT and ANT+
+    bool antplus_flag = false;
+
+    data_t *data;
+
+    // validate buffer: ANT messages are shorter than 150us, i.e. ~140 bits at 1Mbps
+    if (bitbuffer->bits_per_row[0] < 137 || bitbuffer->bits_per_row[0] > 160) {
+        return DECODE_ABORT_LENGTH;
+    }
+
+    // find a data package and extract data buffer
+    bit_offset = bitbuffer_search(bitbuffer, 0, 0, preamble, sizeof(preamble) * 8) + sizeof(preamble) * 8;
+    if (bit_offset + sizeof(b) * 8 > bitbuffer->bits_per_row[0]) { // did not find a big enough package
+        return DECODE_ABORT_LENGTH;
+    }
+
+    // ANT and ANT+ packes have either aa or 55 preamble, depending on the first bit of
+    // the following byte. i.e. 10101010 1xxxxxxx or 01010101 0xxxxxxx
+    // the best way to know which is being used, is to verify which one has a valid CRC
+    // the following code relies on the fact that 55 is aa shifted right
+    bitbuffer_extract_bytes(bitbuffer, 0, bit_offset, b, sizeof(b) * 8); // assume preamble is aa
+
+    // calculate CRC for both alternatives, starting with aa (used by all ANT+ devices)
+    uint16_t crc = crc16(b, 15, 0x1021, 0xffff);
```

Comment:
if crc bits are in the right order it should be crc16ccitt(b, 17) == 0

Reply:
I'm not sure I understand this comment, sorry. Are you saying that it's a property of that crc algorithm that including the crc itself results in a value of 0?

Reply:
Yes, the CRC is the remainder of division, adding that in the remainder is zero.
So generally, if the byte/bit order fits, just compute the CRC over data+remainder and check for `0`.